Various fixes

reneeotten opened the pull request:
hi Guillaume,

in this PR you can find several small fixes that I made while going through the code, they should be self-explanatory from the commit messages. 
A few comments:
- not all the docstrings are finished or perfect, but it's a starting point (there is a docstrings or placeholders now wherever there should be one).
- TODO statements are added to remind us to fix the positional restraint "info" and MC/bootstrap error analysis. I think I know how to fix the latter, but need to study it a bit more to make sure it doesn't break anything else.
- code reformatting with "yapf" has not been done yet, just to make the real changes easier to see here. Once you had the time to go over the PR and merged the stuff you like and agree with, I can do that.

Let me know what you think and feel free to make changes to the PR or tell me what to change before merging it. Thanks!

gbouvignies commented:
Hi Renee,

Thanks for this PR. This looks really good to me. I really appreciate that you looked into the docstrings and naming; I believe there is much to do there...

I didn't realise that `mc` and `bs` were not functional... I'll have to look at why `deepcopy` doesn't work. For `mc`, we can live with a simple `copy` (this works).

Regarding a2ae327, I know this looks like code duplicate, but it is actually a hack to avoid an error message from `lmfit`. For example the example in `examples/CPMG/H1_HN_AntiPhase` would not run. Indeed, when a parameter that depends on other parameters is copied to `params` while the others are not in defined in `params` yet, `param.expr` cannot be processed, which produces error message. Order of addition is not neutral. This is why in my code I first add dummy parameters with the right names to `params` before overwriting them with the right definition. Best solution I have found for the moment, a pretty ugly one too!

If you revert a2ae327, I think it is good to be merged.

reneeotten commented:
okay, I added a few more docstring changes and reverted, as requested, the removal of lmfit code in that one commit. I will continue to look at the fitting code a bit more and see if there is a better way of doing this.

I think, however, that some code in parameters.py is still redundant and I did removed that in f2a5a0d. A did a rebase of the PR and it should merge cleanly.

Regarding the Monte-Carlo and bootstrap analysis, I have to look at this a bit more to convince myself that just using `copy` is indeed sufficient. If you want we can open a separate issue about this to discuss it in more detail.

reneeotten commented:
I had two more commits lying around that I now added to this PR (see above). I'll leave this alone for now until you had the time to look at it and merge (provided that you're happy with it).
Thanks!

gbouvignies commented:
Perfect! Thanks.
